File: performance/src/rxros2_t2.cpp
#include <cstdio>
#include <rxros/rxros2.h>
#include "std_msgs/msg/string.hpp"
#include "ros2_msg/msg/test.hpp"
using namespace rxcpp::operators;
using namespace rxros2::operators;

struct T2: public rxros2::Node {
    T2(): rxros2::Node("T2") {};
    const std::string bulb_data = std::string(1048576, '#');

    static auto age_of(const rclcpp::Time& time) {
        auto curr_time_nano = rclcpp::Clock().now().nanoseconds();
        auto time_nano = time.nanoseconds();
        return curr_time_nano-time_nano;
    }

    static auto mk_test_msg(const int msg_no, const std::string& data) {
        ros2_msg::msg::Test msg;
        msg.msg_no = msg_no;
        msg.data = data;
        msg.time_stamp = rclcpp::Clock().now();
        return msg;
    }

    void run() {
        RCLCPP_INFO(this->get_logger(), "Starting RxCppNode T2:");
        
        rxros2::observable::from_topic<ros2_msg::msg::Test>(this, "/T1T")
            .subscribe ( [this] (const ros2_msg::msg::Test::SharedPtr msg) {
                RCLCPP_INFO(this->get_logger(), "%d,%d,%d", msg->msg_no, msg->data.length(), age_of(msg->time_stamp));});

        rxcpp::observable<>::interval (std::chrono::milliseconds(50))
            | map ([&](int i) { return mk_test_msg(i, bulb_data); })
            | publish_to_topic<ros2_msg::msg::Test> (this, "/T2T");
    }
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto t2 = std::make_shared<T2>();
    t2->start();
    rclcpp::spin(t2);
    rclcpp::shutdown();
    return 0;
}

File: performance/src/rxros2_t2.py
import rclpy
from rxros2_pytest2 import rxros2
from rclpy.time import Time
from rclpy.clock import Clock
from ros2_msg.msg import Test


def age_of(time: Time):
    curr_time_nano = Clock().now().to_msg().nanosec
    time_nano = time.nanosec
    return curr_time_nano-time_nano


def mk_test_msg(msg_no: int, data: str) -> Test:
    msg = Test()
    msg.msg_no = msg_no
    msg.data = data
    msg.time_stamp = Clock().now().to_msg()
    return msg


class T2(rxros2.Node):
    def __init__(self):
        super().__init__("T2")
        self.bulb_data = ''.rjust(1048576, '#')
        self.get_logger().info('Starting RxPyNode T2:')

    def run(self):
        rxros2.from_topic(self, Test, "/T1T").subscribe(
            lambda msg: self.get_logger().info('%d,%d,%d' % (msg.msg_no, len(msg.data), age_of(msg.time_stamp))))

        rxros2.interval(0.05).pipe(
            rxros2.map(lambda msg_no: mk_test_msg(msg_no, self.bulb_data)),
            rxros2.publish_to_topic(self, Test, "/T2T"))


def main(args=None):
    rclpy.init(args=args)
    t2 = T2()
    t2.start()
    rclpy.spin(t2)
    t2.destroy_node()
    rclpy.shutdown()
